### PR TITLE
[UNR-2275] Track references to replicated objects in order to restore them when they move out and in relevance.

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Unlike what the vanilla net driver was doing, we did not track references to replicated actors in replicated properties. It allows restoring pointer in replicated properties to the right actor when they become relevant again.
This is is an addition to how we currently handle unresolved references, since it works only the first time a property is replicated.
This PR essentially add tracking to properties referencing replicated actors, similar to what Unreal was doing, only with FUnrealObjectRef instead of NetworkGUID.
FObjectReferencesMap are extended to also remember the mapped references, and they are wrapped into a utility class, FSpatialObjectRepState.
Then, for each reference, we will add this additional RepState to a map, USpatialReceiver::ObjectRefToRepStateMap.
When an actor is destroyed, we will use this map to move mapped references to unresolved ones, allowing us to resolve them when the actor becomes available again.

#### Tests
https://github.com/spatialos/UnrealGDKTestGyms/pull/11

#### Primary reviewers

---
This change is [Reviewable](https://reviewable.io/reviews/spatialos/unrealgdk/1551)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - The server no longer crashes, when received RPCs are processed recursively.
 - DeploymentLauncher can parse a .pb.json launch configuration.
 - DeploymentLauncher can launch a Simulated Player deployment independently from the target deployment.
+- Track properties containing references to replicated actors, in order to resolve them again in the actor they references moves out and back in relevance.
```
Comment (mironec) [inline CHANGELOG.md]:
```suggestion
- Track properties containing references to replicated actors, in order to resolve them again if the actor they reference moves out and back into relevance.
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -190,10 +310,11 @@ bool USpatialActorChannel::CleanUp(const bool bForDestroy, EChannelCloseReason C
 			Receiver->ClearPendingRPCs(EntityId);
 			Sender->ClearPendingRPCs(EntityId);
 		}
-
-		NetDriver->RemoveActorChannel(EntityId);
+		
+		NetDriver->RemoveActorChannel(EntityId, *this);
 	}
 
+
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 3]
```diff
@@ -83,6 +83,126 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
...
+	UnresolvedRefs.Empty();
+
+	TSet< FUnrealObjectRef > LocalReferencedObj;
+	for (auto& Entry : ReferenceMap)
+	{
+		GatherObjectRef(LocalReferencedObj, UnresolvedRefs, Entry.Value);
+	}
+
+	// TODO : Support references in structures updated by deltas.
+	//UObject* Object = GetObject();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Appreciate you've done some investigation work, but we don't generally comment out blocks of code. Could you please create a tracking ticket, and move the code there. Thanks!

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 4]
```diff
@@ -1143,6 +1264,12 @@ void USpatialActorChannel::RemoveRepNotifiesWithUnresolvedObjs(TArray<UProperty*
 				continue;
 			}
 
+			// Only when there are unresolved refs.
+			if (ObjRef.Value.UnresolvedRefs.Num() == 0)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Why is this check necessary? Doesn't the RefMap contain only unresolved references?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp [hunk 5]
```diff
@@ -9,8 +9,9 @@
 
 DEFINE_LOG_CATEGORY(LogSpatialNetBitReader);
 
-FSpatialNetBitReader::FSpatialNetBitReader(USpatialPackageMapClient* InPackageMap, uint8* Source, int64 CountBits, TSet<FUnrealObjectRef>& InUnresolvedRefs)
+FSpatialNetBitReader::FSpatialNetBitReader(USpatialPackageMapClient* InPackageMap, uint8* Source, int64 CountBits, TSet<FUnrealObjectRef>* InDynamicRefs, TSet<FUnrealObjectRef>& InUnresolvedRefs)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp]: Why make InDynamicRefs a pointer? Seems more straight forward to just require it.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 6]
```diff
@@ -1992,8 +1992,14 @@ void USpatialNetDriver::AddActorChannel(Worker_EntityId EntityId, USpatialActorC
 	EntityToActorChannel.Add(EntityId, Channel);
 }
 
-void USpatialNetDriver::RemoveActorChannel(Worker_EntityId EntityId)
+void USpatialNetDriver::RemoveActorChannel(Worker_EntityId EntityId, USpatialActorChannel& Channel)
 {
+	for (auto& ChannelRefs : Channel.ObjectReferenceMap)
+	{
+		Receiver->CleanupRepStateMap(ChannelRefs.Value);
+		//Receiver->CleanupIncomingRefMap(ChannelRefs.Value, FChannelObjectPair(&Channel, ChannelRefs.Key));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Please remove commented out code

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -1992,8 +1992,14 @@ void USpatialNetDriver::AddActorChannel(Worker_EntityId EntityId, USpatialActorC
 	EntityToActorChannel.Add(EntityId, Channel);
 }
 
-void USpatialNetDriver::RemoveActorChannel(Worker_EntityId EntityId)
+void USpatialNetDriver::RemoveActorChannel(Worker_EntityId EntityId, USpatialActorChannel& Channel)
 {
+	for (auto& ChannelRefs : Channel.ObjectReferenceMap)
+	{
+		Receiver->CleanupRepStateMap(ChannelRefs.Value);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Any reason we do this here, and not when the channel is closed/shutdown?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 8]
```diff
@@ -742,14 +742,39 @@ void USpatialReceiver::RemoveActor(Worker_EntityId EntityId)
 		return;
 	}
 
-	// If the entity is to be deleted after having been torn off, ignore the request (but clean up the channel if it has not been cleaned up already).
-	if (Actor->GetTearOff())
+	
+	if (USpatialActorChannel* ActorChannel = NetDriver->GetActorChannelByEntityId(EntityId))
 	{
-		if (USpatialActorChannel* ActorChannel = NetDriver->GetActorChannelByEntityId(EntityId))
+		for (int32 i = 0; i < ActorChannel->CreateSubObjects.Num(); i++)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Best to use a range-based for loop here

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 9]
```diff
@@ -742,14 +742,39 @@ void USpatialReceiver::RemoveActor(Worker_EntityId EntityId)
...
+		{
+			if (UObject *SubObject = ActorChannel->CreateSubObjects[i])
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
			if (UObject* SubObject = ActorChannel->CreateSubObjects[i])
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 10]
```diff
@@ -742,14 +742,39 @@ void USpatialReceiver::RemoveActor(Worker_EntityId EntityId)
...
+			{
+				FUnrealObjectRef ObjectRef = FUnrealObjectRef::FromObjectPtr(SubObject, Cast<USpatialPackageMapClient>(PackageMap));
+				// Unmap this object so we can remap it if it becomes relevant again in the future
+				MoveMappedObjectToUnmapped(ObjectRef);
+			}
+		}
+
+		if (Actor != nullptr)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Actor's already been checked above.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 11]
```diff
@@ -2114,3 +2184,43 @@ void USpatialReceiver::PeriodicallyProcessIncomingRPCs()
...
+		return TEXT("<unknown>");
+	}
+}
+
+void USpatialReceiver::CleanupRepStateMap(FSpatialObjectRepState& RepState)
+{
+	for (const FUnrealObjectRef& Ref : RepState.ReferencedObj)
+	{
+		TSet<FSpatialObjectRepState*>* RepStatesWithMappedRef = ObjectRefToRepStateMap.Find(Ref);
+		if (ensureMsgf(RepStatesWithMappedRef, TEXT("Ref to entity %i on object %s is missing its referenced entry in the Ref/RepState map"), Ref.Entity, *GetObjectNameFromRepState(RepState)))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Please use %lld for all your entity id references

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 12]
```diff
@@ -83,6 +83,126 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
...
+	if (CurReferences.Array)
+	{
+		for (auto const& Entry : *CurReferences.Array)
+		{
+			GatherObjectRef(OutReferenced, OutUnresolved, Entry.Value);
+		}
+	}
+
+	OutUnresolved.Append(CurReferences.UnresolvedRefs);
+	OutReferenced.Append(CurReferences.UnresolvedRefs);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Why add the UnresolveRefs to the referenced map?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 13]
```diff
@@ -1475,29 +1558,31 @@ void USpatialReceiver::ReceiveCommandResponse(const Worker_CommandResponseOp& Op
 	}
 }
 
-void USpatialReceiver::ApplyComponentUpdate(const Worker_ComponentUpdate& ComponentUpdate, UObject* TargetObject, USpatialActorChannel* Channel, bool bIsHandover)
+void USpatialReceiver::ApplyComponentUpdate(const Worker_ComponentUpdate& ComponentUpdate, UObject& TargetObject, USpatialActorChannel& Channel, bool bIsHandover)
 {
-	FChannelObjectPair ChannelObjectPair(Channel, TargetObject);
+	//FChannelObjectPair ChannelObjectPair(Channel, TargetObject);
+	TWeakObjectPtr<UObject> ObjectPtr = MakeWeakObjectPtr(&TargetObject);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: This doesn't seem to be used

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 14]
```diff
@@ -19,6 +20,96 @@
...
+	USpatialActorChannel* Channel;
+	const FClassInfo* Info;
+	TWeakObjectPtr<UObject> Subobject;
+
+	TSet<Worker_ComponentId> PendingAuthorityDelegations;
+};
+
+class FSpatialObjectRepState;
+
+using FObjectToRepStateMap = TMap < FUnrealObjectRef, TSet<FSpatialObjectRepState*> >;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h]: Move to SpatialCommonTypes.h

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 15]
```diff
@@ -228,10 +179,11 @@ class USpatialReceiver : public UObject
 
 	FTimerManager* TimerManager;
 
-	// TODO: Figure out how to remove entries when Channel/Actor gets deleted - UNR:100
-	TMap<FChannelObjectPair, FObjectReferencesMap> UnresolvedRefsMap;
-	TArray<TPair<UObject*, FUnrealObjectRef>> ResolvedObjectQueue;
+	struct RepStateUpdateHelper;
+	TMap<FUnrealObjectRef, TSet<FSpatialObjectRepState*>> ObjectRefToRepStateMap;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h]: Please add a helper comment here

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 16]
```diff
@@ -83,6 +83,126 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
...
+			RepStateMap.FindOrAdd(Ref).Add(this);
+		}
+	}
+
+	// Remove any guids that we were previously tracking but no longer should
+	for (const FUnrealObjectRef& Ref : ReferencedObj)
+	{
+		if (!LocalReferencedObj.Contains(Ref))
+		{
+			TSet< FSpatialObjectRepState* >& RepStatesWithMappedRef = RepStateMap.FindChecked(Ref);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Mapped/Unmapped isn't relevant now

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 17]
```diff
@@ -1847,29 +1917,33 @@ void USpatialReceiver::ResolveIncomingOperations(UObject* Object, const FUnrealO
...
+		FChannelObjectPair ChannelObjectPair = RepState->GetChannelObjectPair();
+		USpatialActorChannel* DependentChannel = ChannelObjectPair.Key.Get();
+		
+		if (!RepState->GetChannelObjectPair().Value.IsValid())
 		{
-			UnresolvedRefsMap.Remove(ChannelObjectPair);
+			if (DependentChannel->ObjectReferenceMap.Find(ChannelObjectPair.Value))
+			{
+				DependentChannel->ObjectReferenceMap.Remove(ChannelObjectPair.Value);
+			}
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Add cleanup code to SpatialActorChannel::ReplicateActor where it looks for deleted subobjects.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 18]
```diff
@@ -1956,12 +2025,8 @@ void USpatialReceiver::ResolveObjectReferences(FRepLayout& RepLayout, UObject* R
...
-			ResolveObjectReferences(RepLayout, ReplicatedObject, *ObjectReferences.Array, bIsHandover ? nullptr : (uint8*)StoredArray->GetData(), (uint8*)Array->GetData(), NewMaxOffset, RepNotifies, bOutSomeObjectsWereMapped, bArrayHasUnresolved);
-			if (!bArrayHasUnresolved)
-			{
-				It.RemoveCurrent();
-			}
-			else
+			ResolveObjectReferences(RepLayout, ReplicatedObject, RepState, *ObjectReferences.Array, bIsHandover ? nullptr : (uint8*)StoredArray->GetData(), (uint8*)Array->GetData(), NewMaxOffset, RepNotifies, bOutSomeObjectsWereMapped, bArrayHasUnresolved);
+			if (bArrayHasUnresolved)
 			{
 				bOutStillHasUnresolved = true;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Perhaps we don't need this anymore?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 7 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: This is called from the channel's Cleanup and Close methods. Is there a better place to do this cleanup ?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 12 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Explained in comment

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 4 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Now it contains both.

File: CHANGELOG.md [hunk 19]
```diff
@@ -6,6 +6,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased-`x.y.z`] - 2020-xx-xx
 
+- Minor spelling fix to connection log message
+- The GDK now uses SpatialOS `14.3.0`.
```
Comment (joshuahuburn) [inline CHANGELOG.md]: This appears to be a duplication of what's below in `### Features:`

**Pull request merged**